Mounted Bow and other vehicle changes

mugg91 opened the pull request:
## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.
 
## Listing (Includes all commits)

[changed] Mounted Bow cannot fall through small gaps anymore.
[changed] Mounted Bow can now collide with platforms, doors, etc. but not with larger blobs.
[changed] Mounted Bow bought from Trader now can be converted by picking it up.
[changed] Airship and Bomber now show healthbar, can be converted and don't decay if left alone.
[changed] Mounted Bow which is attached to another vehicle can now always be detached via button.
[changed] Catapult and Ballista which is attached to another vehicle can now be detached if touching ground or when airborne.
[changed] Vehicles' ammo counter is not hidden in single player anymore.
[changed] You will detach from or let go of vehicles that aren't from your team.
[fixed] Mounted Bow doesn't look bugged when facing the other direction.
[fixed] Vehicles attached to other vehicles don't look bugged when facing the other direction.
[fixed] Bow on Warboat now gets converted along with the Warboat.
[fixed] Converting a Vehicle that is attached to another vehicle will detach instead of converting the other vehicle.

**Team Checking Commit:** 

[changed] You can use Storage and Saw if it is in team 255.
[changed] You can change class at Tent from team 255.
[changed] Longboat doesn't Decay if left alone.
[changed] You can interact with vehicles if they are in team 255.
[fixed] Inconsistent team checking in Hall.as, HallCommon.as, MigrantBrain.as, SeatHop.as, SeatsGUI.as, VehicleConvert.as, RedBarrier.as.
[fixed] That Migrants converted to your team didn't receive your color.

**TODO:**

[Still unfixed...] Gunner in non-neutral team shooting from a neutral vehicle will cause damage to the vehicle.

**Fixed in a different PR:**

[fixed] Shooting from a Mounted Bow that is in a Catapult's mag will cause the Catapult to fly away. - #1311 
[fixed] Mounted Bow shows a wrong "load item" button after it was put in a Catapult's mag. - #1311 

## Description (Only initial commits shown)

- Mounted Bow doesn't fall through small gaps anymore. Since player can attach himself to it, the player could also fall through the gap with it, causing unwanted behavior or wall clipping. This was done by changing verticesXY.
- Bomber and Airship now show health bar. Instead of decaying if an enemy is near, it can be converted to the enemy now. As for the health bar, I think all big vehicles should show it.
- Mounted Bow/Catapult/Ballista no longer look bugged when they were attached to Bomber/Airship/Warship and then changing facing direction. Mounted Bow had this problem even when not attached.
- Mounted Bow can now always be detached from another vehicle. It was not possible before. The Mounted Bow on Warboat is unaffected and cannot be detached.
- Catapult and Ballista can now be detached if the other vehicle is either "touching ground" or "not touching ground but not touching water either". Therefore, Catapult and Ballista cannot be detached from a Warboat that is riding on water without ground. But it can be detached from Airship and Bomber as long as they are not in water. I think this feels the most practical. But of course feel free to say if you have a different opinion. 
- Mounted Bow doesn't fall through doors, platforms, etc.
- All vehicles' ammo counts are no longer hidden in single player.
- Mounted Bow can be picked up and thereby be converted to your team. Mounted Bow that has been there from the start cannot be picked up and will stay neutral and stationary, possible to be used by all teams.

## Steps to Test or Reproduce

- Place a small gap and !mounted_bow. Notice it doesn't fall into the gap anymore.
- !bomber and !airship. Notice they don't get destroyed if you go away. Notice they have health bar. 
!team 1 , notice they can be converted to Red Team.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. Shooting diagonally with it. Detach yourself and pick it up. Face the other direction. Alternatively, you can load it on a Catapult and ride the other direction. Notice it doesn't look bugged anymore.
- Go to Sandbox, !catapult, !bomber. The Catapult should be attached to the Bomber. Get seated in the Bomber and make it face the other direction. Notice the Catapult doesn't look bugged anymore.
- !bomber (or !airship or !warboat), !mounted_bow. Notice if you show buttons, there is a "Detach" button.
- !bomber, !catapult. Use the bomber to fly. Notice you can "Detach" the Catapult while in the air.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. !tent, change class to builder, !wood, build some doors and platforms. Throw the Mounted Bow into the doors and platforms. Notice that it collides with them now.
- Go to Save the Princess, !tradingpost, !coins, buy a Mounted Bow. Throw it away. !team 1. Notice that it will get converted to Red Team if you pick it up. If you go to Sandbox, /loadmap "Biurza_Black_Dog", you can notice that the Mounted Bow in the middle does not get converted to any team when attaching yourself to it.
- Go to any single player mode and use a catapult or a mounted bow. Notice that they show the ammo count now.

mugg91 commented:
While trying to fix "Mounted Bow on Catapult bug", I found this:
1) Catapult.as detaches the MountedBow upon use.
2) In MountedBow.as OnCommand(), blob will be an arrow when used in any other circumstance but will be set to catapult's id when it is loaded on the Catapult. 
params.read_netid() will somehow return the catapult's id.
I'm not sure if I will be able to fix this but I will keep trying.

mugg91 commented:
I'm done with this PR, so setting this to "READY". I was not able to fix these two issues, though:

- [issue #1247 - Mounted Bow on Catapult bug](https://github.com/transhumandesign/kag-base/issues/1247)
- Mounted Bow erroneously showing a "load item on this" button after it was loaded on a Catapult.

mugg91 commented:
I added another change.
When a vehicle is converted from your team to enemy team, you will be detached from it or let go of it.

Reproduction Steps:

- Be on Blue Team and hold a Blue Dinghy.
- Go to two Red Builders, crouch, wait for your Blue Dinghy to be converted to Red.
- Instead of holding a Red Dinghy, you will be detached from it.

----

- Be on Blue Team and sit in a Blue Catapult.
- Have two Red Builders nearby so your Catapult is converted from Blue to Red.
- Instead of sitting in a Red Catapult, you will be detached from it.

mugg91 commented on Entities/Common/Scripts/SetTeamToCarrier.as:
```diff
@@ -8,7 +8,16 @@ void onAttach(CBlob@ this, CBlob@ attached, AttachmentPoint@ attachedPoint)
+	
+	CBlob@ carried = attached.getCarriedBlob();
+	
+	if (carried !is null)
+	{
+		if (carried.getName() == this.getName())
```
`carried.getName()` will cause an exception if carried is null.

```
CBlob@ testblob;
if (testblob !is null || testblob.getName() == "test") // This causes an exception
{ print("hi"); } 
```

```
CBlob@ testblob;
if (testblob !is null)
	if (testblob.getName() == "test") // This doesn't
	{
		print("hi");
	}
```

mugg91 commented on Entities/Structures/MountedBow/MountedBow.cfg:
```diff
@@ -89,15 +89,15 @@ $inventory_name                            = Load
-										 GenericHit.as;
-										 CollidesOnlyWithLarger.as;	
-										 AutoGrabFromGround.as;						 
+					 Seats.as;      # set this first
+					 IsFlammable.as;
+					 WoodVehicleDamages.as;
+					 Vehicle.as;
+					 MountedBow.as;
+					 Wooden.as;
+					 GenericHit.as;
+					 SetTeamToCarrier.as;
+					 AutoGrabFromGround.as;	
```
It is unclear how it is supposed to look, since it looks different in my editor vs. github.

mugg91 commented on Entities/Vehicles/Common/VehicleCommon.as:
```diff
@@ -551,15 +547,36 @@ void Vehicle_StandardControls(CBlob@ this, VehicleInfo@ v)
-				{
+				if 	(occupied.isMyPlayer() && ap.isKeyJustPressed(key_up)
+					|| (this.getTeamNum() != occupied.getTeamNum() 
+						&& (this.getTeamNum() >= 0 && this.getTeamNum() <= 7)
+						&& (occupied.getName() == "knight" || occupied.getName() == "archer" || occupied.getName() == "builder")))
```
"knight", "archer" and "builder is hard coded in a lot more places. 
The ItemLimits script I made will also check for team number like this.
How to avoid this?
Although neutral blobs are usually set to -1 or 255, you can change to teams 8~254 via `!team (number)` in Sandbox. Is there a way to check if a blob is in grey team that doesn't involve checking team number?

eps0003 replied:
> "knight", "archer" and "builder is hard coded in a lot more places.
> How to avoid this?

You should always endeavour to leave code in a better state than you found it. A potential solution would be to check for `"player"` tag, but migrants also have this tag which could cause rare side effects.

> Although neutral blobs are usually set to -1 or 255, you can change to teams 8~254 via !team (number) in Sandbox. Is there a way to check if a blob is in grey team that doesn't involve checking team number?

There is nothing special with grey teams. They are only grey because there aren't enough colours to cover all team numbers. The only team that should be treated differently from the rest is spectator team (`getRules().getSpectatorTeamNum()`). -1 == 255 because team is u8 which wraps negative numbers around to the maximum.

mugg91 replied:
I can change to `occupied.hasTag("player")`. I don't see a problem since migrants are never attached to vehicles.

With the team check, this kind of check is in many places. I guess teams 0~254 should be treated as player teams and team 255 as a neutral team for animals and stuff that can be used by anyone (vehicles, doors).
I have seen places where the team check says "anything that is higher than team 100" or "anything that is higher than team 10". Those checks seem arbitrary and inconsistent. Up until now, I tried to be a more precise by handling teams 0~7 as player teams and anything else as neutral team.

I can change that so that 0~254 are player teams instead. And 255 is a neutral team.
Is that what should be done?

eps0003 replied:
> I can change that so that 0-254 are player teams instead. And 255 is a neutral team. Is that what should be done?

Yes, this sounds like the most appropriate way of dealing with teams. Please remember to check for `getRules().getSpectatorTeamNum()` rather than specifically team 255.

mugg91 replied:
Sorry if it was unclear. I was not talking about spectator team.
I'm still talking about **teams that are currently on the field playing**. Of those, 0 to 254 should be handled as teams that players could get assigned to. Players can also get assigned to team 255, but behavior will be different.

eps0003 replied:
Hold up... why is there even a team or player check here? If the blob is already sitting on the vehicle, they must've been the correct team to get on it in the first place. And there is no need to check if it's a player since non-player blobs will never be pressing `key_up` to get out. Even if they somehow did, there is no harm in allowing them to detach from the vehicle like players do.

Looking at SeatHop.as line 32, there already appears to be a check to allow neutral teams to sit in vehicles.

mugg91 replied:
It is so that when you are sitting in your vehicle and it is getting converted to the enemy, you are kicked out.
The player check is so that enemy crates and enemy Dinghies can be loaded/can stay in the catapult's mag.

I'm not working on "neutral players doing stuff" but I'm working on "any players using neutral stuff".

About `getRules().getSpectatorTeamNum()`, I can use it, but the Map loader will set blobs to -1 so I'm afraid of mismatch errors.

eps0003 replied:
Team is a `u8` which means -1 wraps around to 255 (the maximum). `getRules().getSpectatorTeamNum()` returns 255 so it should match correctly.

mugg91 replied:
In Sandbox mode, `getRules().getSpectatorTeamNum()` returns 200 for me.
I'm reworking team checks and adding another commit later.

mugg91 commented on Entities/Vehicles/Common/VehicleCommon.as:
```diff
@@ -551,15 +547,36 @@ void Vehicle_StandardControls(CBlob@ this, VehicleInfo@ v)
+			{
+				CBlob@ carrier = pickup.getOccupied();
+				
+				if (carrier !is null)
+				{
+					if (this.getTeamNum() != carrier.getTeamNum())
```
Please read my other response.

mugg91 commented:
Fixed some vehicle conversion bugs:
- Bow on Warboat will get converted along with the Warboat now.
- When a vehicle which is attached to a carrier vehicle gets converted, it will detach now instead of erroneously converting the carrier vehicle.

mugg91 commented:
The merge commit and the closing/re-opening of this PR are misclicks, please do not mind.
The thing to note is commit [6dfb653](https://github.com/transhumandesign/kag-base/pull/1251/commits/6dfb653b619ae6174516ceec9ac8d1b36ef08997).

- Added TeamChecking.as. 
This was done because there are too many inconsistent checks and [eps0003](https://github.com/eps0003) urged me to find better ways of checking teams.
Treat 0-254 as player teams and 255 as neutral team.
This was done in this PR because a lot of team checking was already done here.
The functions in TeamChecking.as are shared because HallCommon.as calls from it in a shared function, otherwise it would cause an error when loading a new map with a Hall in it.
- Storage: Can be used if it is in your team or in team 255.
- Hall: Minimap icon differentiates between team 0-254 and 255 instead of 0-9 and 10-254.
- Hall: Only show buttons when the hall is yours and you are nearby.
- Hall: Migrants converted to your team will receive your color.
- Vehicles: Don't receive coins for damaging or destroying vehicles in team 255.
- Vehicles: Can sit in, ammo-load, item-load, detach, look in inventory of, flip vehicles in your team or in team 255.
- Vehicles: Ignore gunner's team and make projectile be team 255 if the vehicle is team 255, so vehicle doesn't hit itself.
- Tent: Can change class if tent is in team 255.
  (I made a different PR which deals with resupplies from enemy structures. Maybe worth considering to let neutral bases give resupplies.)
- Longboat: Removed DecayIfLeftAlone.as (It makes no sense that it could be converted but also get auto-destroyed when only the enemy is there)
- Red Barrier: Only team 255 can pass, instead of teams 100-255.

mugg91 commented:
Ok, I may have gone a little overboard with this latest team checking commit.
But I hope you understand why I did it.
I guess it makes more sense to split it into its own PR.
I just had figured, it may save trouble down the line to reduce workload of whoever has to merge all these things later.
Let me know what should be done.

Vam-Jam commented:
I'm going to close this PR as there's a lot of merge issues + It's a big PR that tries to do too much. 

I would suggest you split it up into separate PR's. E.g. Vehicles into a vehicle PR, 255 team changes into a 255 team pr etc.

mugg91 commented:
> I'm going to close this PR as there's a lot of merge issues + It's a big PR that tries to do too much.
> 
> I would suggest you split it up into separate PR's. E.g. Vehicles into a vehicle PR, 255 team changes into a 255 team pr etc.

I opened separate issues for each problem, except things related to team checking or getting resupplies from enemy buildings.
